protocol/packages/currencies/build/currencies_tree.rs

```diff
@@ -0,0 +1,130 @@
+use std::{
+    borrow::Borrow,
+    collections::{
+        btree_map::{self, BTreeMap},
+        btree_set::{self, BTreeSet},
+    },
+};
+
+use anyhow::{Context as _, Result};
+
+use topology::{swap_pairs::PairTargets, Topology};
+
+use crate::protocol::Protocol;
+
+pub(crate) struct CurrenciesTree<'parents_of, 'parent, 'children_of, 'child> {
+    parents: BTreeMap<&'parents_of str, BTreeSet<&'parent str>>,
+    children: BTreeMap<&'children_of str, BTreeSet<&'child str>>,
+}
+
+impl<'topology> CurrenciesTree<'topology, 'topology, 'topology, 'topology> {
+    const EMPTY: Self = Self {
+        parents: BTreeMap::<_, BTreeSet<_>>::new(),
+        children: BTreeMap::<_, BTreeSet<_>>::new(),
```

Comment:
I think `Parents` and `Children` is what you wanted to use here, instead of `BtreeSet`?

Reply:
Sadly no, I thought the same thing when I extracted it, but `Parents` and `Children` actually hold references. I should be able to rework them a bit actually and use them here.
Also, I should be able to elide the generics here, as the types are well-know.

Reply:
without claiming that I understand the entire mechanics here, I would suggest reviewing if the ownership of the tickets could be brought here and allow the usage of Parents and Children.